packages/queued-request-controller/src/QueuedRequestMiddleware.ts

```diff
@@ -0,0 +1,140 @@
+    'eth_sendTransaction',
+    'wallet_watchAsset',
+    'wallet_switchEthereumChain',
+    'eth_signTypedData_v4',
+  ];
+
+  return !confirmationMethods.includes(method);
+};
+
+export const createQueuedRequestMiddleware = (
+  messenger: ControllerMessenger<
+    | QueuedRequestControllerEnqueueRequestAction
+    | NetworkControllerGetStateAction
+    | NetworkControllerSetActiveNetworkAction
+    | NetworkControllerSetProviderTypeAction
+    | NetworkControllerGetNetworkClientByIdAction
+    | NetworkControllerFindNetworkClientIdByChainIdAction
+    | SelectedNetworkControllerSetNetworkClientIdForDomainAction
+    | AddApprovalRequest,
+    never
+  >,
+  useRequestQueue: () => boolean,
+): JsonRpcMiddleware<any, any> => {
+  return createAsyncMiddleware(async (req: any, res: any, next) => {
+    if (!useRequestQueue() || isConfirmationMethod(req.method)) {
+      next();
+      return;
+    }
+
+    const networkClientIdForRequest = req.networkClientId as NetworkClientId;
+
+    await messenger.call(
+      QueuedRequestControllerActionTypes.enqueueRequest,
+      async () => {
+        if (req.method === 'wallet_switchEthereumChain') {
+          // eslint-disable-next-line n/callback-return
+          return next();
+        }
+
+        const isBuiltIn = isNetworkType(networkClientIdForRequest);
+        let networkConfigurationForRequest;
+        if (isBuiltIn) {
+          const builtIn = BUILT_IN_NETWORKS[networkClientIdForRequest];
+          if (builtIn.chainId) {
+            networkConfigurationForRequest = builtIn;
+          }
+        }
+
+        networkConfigurationForRequest =
+          networkConfigurationForRequest ||
+          messenger.call(
+            'NetworkController:getNetworkClientById',
+            networkClientIdForRequest,
+          ).configuration;
+
+        const currentProviderConfig = messenger.call(
+          'NetworkController:getState',
+        ).providerConfig;
+
+        const currentChainId = currentProviderConfig.chainId;
+
+        if (currentChainId === networkConfigurationForRequest.chainId) {
+          // eslint-disable-next-line n/callback-return
+          return next();
+        }
+
+        // if is switch eth chain call
+        // clear request queue when the switch ethereum chain call completes (success or fail)
+        // This is because a dapp-requested switch ethereum chain invalidates any requests they've made after this switch, since we dont know if they were expecting the chain after the switch or before.
+        // with the queue batching approach, this would mean clearing any batch for that origin (batches being per-origin.)
+        const requestData = {
+          toNetworkConfiguration: networkConfigurationForRequest,
+          fromNetworkConfiguration: currentProviderConfig,
+        };
+
+        try {
+          const approvedRequestData = await messenger.call(
+            'ApprovalController:addRequest',
+            {
+              origin,
```

Comment:
Where is this `origin` variable coming from? I can't find it anywhere in this file. I assume the intention is:

```suggestion
              origin: req.origin,
```

or with my proposed changes:

```suggestion
              origin: req.params?.origin ?? '',
```

Reply:
Yeah, a curious thing about the `request` object in middleware is that it's mutable, and we ~exploit~ take advantage of this quite liberally. In this case, `origin` is being added by the middleware in the clients. In the extension, it's added [here](https://github.com/MetaMask/metamask-extension/blob/f0d81dcc514c92be1f5192b8af701a9ade35fcdf/app/scripts/lib/createOriginMiddleware.js), and in the mobile app, it's added [here](https://github.com/MetaMask/metamask-mobile/blob/c9a1b3e0d87e8344d6c17cd7a28cf8945d8b0c8b/app/util/middlewares.js#L29).

Reply:
Yeah not too sure how that wasn't giving errors =\